utils.go

```diff
@@ -81,3 +83,120 @@ func toNRGBA(img image.Image) *image.NRGBA {
 	}
 	return Clone(img)
 }
+
+// nrgbaToHSL converts NRGBA to HSL.
+func nrgbaToHSL(c color.NRGBA) (float64, float64, float64) {
+	var h, s, l float64
+
+	r := float64(c.R) / float64(255)
+	g := float64(c.G) / float64(255)
+	b := float64(c.B) / float64(255)
+
+	min := math.Min(math.Min(r, g), b)
+	max := math.Max(math.Max(r, g), b)
+
+	l = (max + min) / 2
+
+	if min == max {
+		s = 0
+		h = 0
+	} else {
+		if l < 0.5 {
+			s = (max - min) / (max + min)
+		} else {
+			s = (max - min) / (2.0 - max - min)
+		}
+
+		if max == r {
+			h = (g - b) / (max - min)
+		} else if max == g {
+			h = 2.0 + (b-r)/(max-min)
+		} else {
+			h = 4.0 + (r-g)/(max-min)
+		}
+
+		h *= 60
+
+		if h < 0 {
+			h += 360
+		}
+	}
+
+	return h, s, l
+}
+
+// hslToNRGBA converts HSL to NRGBA with A=1.
+func hslToNRGBA(h, s, l float64) color.NRGBA {
+	if s == 0 {
+		c := uint8(l * 255)
+		return color.NRGBA{R: c, G: c, B: c, A: 1}
+	}
+
+	var r, g, b float64
+	var t1, t2, tr, tg, tb float64
+
+	if l < 0.5 {
+		t1 = l * (1.0 + s)
+	} else {
+		t1 = l + s - l*s
+	}
+
+	t2 = 2*l - t1
+	h = h / 360
+	tr = h + 1.0/3.0
+	tg = h
+	tb = h - 1.0/3.0
+
+	if tr < 0 {
+		tr++
+	} else if tr > 1 {
+		tr--
+	}
+
+	if tg < 0 {
+		tg++
+	} else if tg > 1 {
+		tg--
+	}
+
+	if tb < 0 {
+		tb++
+	} else if tb > 1 {
+		tb--
+	}
+
+	// Red
+	if 6*tr < 1 {
+		r = t2 + (t1-t2)*6*tr
+	} else if 2*tr < 1 {
+		r = t1
+	} else if 3*tr < 2 {
+		r = t2 + (t1-t2)*(2.0/3.0-tr)*6
+	} else {
+		r = t2
+	}
+
+	// Green
+	if 6*tg < 1 {
+		g = t2 + (t1-t2)*6*tg
+	} else if 2*tg < 1 {
+		g = t1
+	} else if 3*tg < 2 {
+		g = t2 + (t1-t2)*(2.0/3.0-tg)*6
+	} else {
+		g = t2
+	}
+
+	// Blue
+	if 6*tb < 1 {
+		b = t2 + (t1-t2)*6*tb
+	} else if 2*tb < 1 {
+		b = t1
+	} else if 3*tb < 2 {
+		b = t2 + (t1-t2)*(2.0/3.0-tb)*6
+	} else {
+		b = t2
+	}
+
+	return color.NRGBA{R: uint8(r * 255), G: uint8(g * 255), B: uint8(b * 255)}
```

Comment:
I think rounding the values will give more accurate result. E.g. `clamp(r * 255)` instead of `uint8(r * 255)` etc.